src/NuGet.Clients/NuGet.CommandLine/Commands/SearchCommand.cs

```diff
@@ -0,0 +1,208 @@
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using System.Threading;
+using Newtonsoft.Json;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Frameworks;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Packaging.Licenses;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+
+namespace NuGet.CommandLine
+{
+
+    [Command(typeof(NuGetCommand), "search", "SearchCommandDescription",
+            UsageSummaryResourceName = "SearchCommandUsageSummary", UsageExampleResourceName = "SearchCommandUsageExamples")]
+    public class SearchCommand : Command
+    {
+        private readonly List<string> _sources = new List<string>();
+
+        [Option(typeof(NuGetCommand), "SearchCommandSourceDescription")]
+        public ICollection<string> Source
+        {
+            get { return _sources; }
+        }
+
+        // -Source "https://api.nuget.org/v3/index.json"
+
+
+        [Option(typeof(NuGetCommand), "SearchCommandPreRelease")]
+        public bool PreRelease { get; set; } = false;
+
+
+        [Option(typeof(NuGetCommand), "SearchCommandTake")]
+        public int Take { get; set; } = 20;
+
+
+        private IList<PackageSource> GetEndpointsAsync()
+        {
+            var configurationSources = SourceProvider.LoadPackageSources()
+                .Where(p => p.IsEnabled)
+                .ToList();
+
+            IList<PackageSource> packageSources;
+            if (Source.Count > 0)
+            {
+                packageSources = Source
+                    .Select(s => PackageSourceProviderExtensions.ResolveSource(configurationSources, s))
+                    .ToList();
+            }
+            else
+            {
+                packageSources = configurationSources;
+            }
+            return packageSources;
+        }
+
+        public override async Task ExecuteCommandAsync()
+        {
+            ILogger logger = Console;
+            CancellationToken cancellationToken = CancellationToken.None;
+
+            SearchFilter searchFilter = new SearchFilter(includePrerelease: PreRelease);
+
+            var listEndpoints = GetEndpointsAsync();
+
+            foreach (var source in listEndpoints)
+            {
+                var target = source.Source;
+                var name = source.Name;
+
+                SourceRepository repository = Repository.Factory.GetCoreV3(target);
+                PackageSearchResource resource = await repository.GetResourceAsync<PackageSearchResource>();
+
+                if (resource is null)
+                {
+                    continue;
+                }
+
+                IEnumerable<IPackageSearchMetadata> results = await resource.SearchAsync(
+                    string.Join("+", Arguments).Trim().Replace(' ', '+'),
```

Comment:
Why do join using `"+"` instead of a space? Consider adding a comment here to clarify

Reply:
I added a comment on that line explaining that we join the args using '+' so that it's formatted correctly for the search query URL: "The arguments (query strings) are joined with '+' so that the queries are passed to the search query url with the correct format (eg. "logging extensions" --> "logging+extensions")"

Reply:
I changed the code to join the arguments with spaces rather than '+' to maintain consistency with PM UI.

Reply:
I believe this is the code that builds the Query String, and I didn't see anything converting args to +.
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Resources/PackageSearchResourceV3.cs#L117

I wonder if the HttpRequestMessage automatically converts query strings as needed? https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpRequestMessageFactory.cs#L101

Maybe try this, and see if you can eliminate the string manipulation you have, here. If you find you must keep it, I'd think we should move this to one of the helper classes, either a Utility or Extensions class we already have somewhere.

Reply:
The PM UI does not encode search query strings properly (to the point where it may be a security issue). See: https://github.com/NuGet/Home/issues/9043

Reply:
Would the best step for me right now be to simply join query strings using spaces, and leave the fix to the lower level code (that actually constructs the url) for a later time then?